<a href="https://colab.research.google.com/github/Avipsa1/UPPP275-Notebooks/blob/main/Geocoding.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Geocoding

What is geocoding?

Geocoding is to converti addresses into coordinates (or vice versa), which is a really common GIS task. For example, we might want to know the longitude and latitude from a business address. Fortunately, there are convinient libraries in python that make the geocoding task really easy.



<br></br>
One of the libraries that can do the geocoding for us is `geopy` (https://geopy.readthedocs.io/en/stable/#module-geopy.geocoders) that makes it easy to locate the coordinates of addresses, cities, countries, and landmarks across the globe using third-party geocoders and other data sources. You can think about `geopy` as a "wrapper"; however, what's inside is geocoder services, and many of them are not FREE! (also can be very expensive if you have large queries). Some of the providers are:

* ArcGIS online geocoding service (https://geocode.arcgis.com).
* Google Geocoding API (https://developers.google.com/maps/documentation/geocoding/start).
* OpenStreetMap (Nominatim) (https://nominatim.openstreetmap.org).




<br></br>
Great thing about `geopy` is that it is integrated in `geopandas`, which makes it very friendly (native) to use. Now let's give it a try.





In [ ]:
my_address = "300 Campus Drive, Irvine, California"

<br></br>
Above is my apartment address, is someone living close to me?
<br></br>
And next, let's try to convert this address to coordinates. The command is: 


<br></br>
`geopandas.tools.geocode.eocode(my_address, provider='nominatim')`
    
where we use the OpenStreetMap's Nominatim's geocoding service.


In [ ]:
!pip install geopandas
from geopandas.tools import geocode
# Geocode addresses with Nominatim backend
coords = geocode(my_address, provider='nominatim')

/usr/local/lib/python3.7/dist-packages/geopy/geocoders/osm.py:143: UserWarning: Using Nominatim with the default "geopy/1.17.0" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  UserWarning


In [ ]:
coords

geometry  \
0  POINT (-117.85272 33.66512)   

                                             address  
0  Campus Drive, Irvine Business Complex, Irvine,...



<br></br>
What we get is a GeoDataFrame with only one row that contains my address and the coordinates, which seems to be quite accurate. (you can validate it on Google Maps)

<br></br>
`POINT (-117.85272 33.66512)`



<br></br>
What if we have multiple addresses? We just need to give the `geocode` function a list of addresses. 
<br></br>
For example, if we have two locations: Campus Drive and UCI Anteater pavilion.



In [ ]:
my_address_list = ["300 Campus Drive, Irvine, California", 
                   "200 Harvard Avenue, Irvine, California"]

coords_list = geocode(my_address_list, provider='nominatim')

/usr/local/lib/python3.7/dist-packages/geopy/geocoders/osm.py:143: UserWarning: Using Nominatim with the default "geopy/1.17.0" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  UserWarning


In [ ]:
coords_list

geometry  \
0  POINT (-117.85272 33.66512)   
1  POINT (-117.83277 33.67103)   

                                             address  
0  Campus Drive, Irvine Business Complex, Irvine,...  
1  Harvard Avenue, Rancho San Joaquin, Irvine, Or...

That being said, if we have a Excel table of addresses and needed be geocoded, we can do this in a batch.

Let's try to plot those two locations on a map using the coordinates we obtained from the geocoder (I will talk about how to make a map like below  in more details in the Module 6).

In [ ]:
import folium
%matplotlib inline
irvine = folium.Map(
    location=[33.6651, -117.8527],
    tiles='Stamen Toner',
    zoom_start=11
)
folium.Marker([coords_list.geometry[1].y, coords_list.geometry[1].x], popup='<i>Michelson Drive</i>').add_to(irvine)
folium.Marker([coords_list.geometry[0].y, coords_list.geometry[0].x], popup='<i>Jamboree Road</i>').add_to(irvine)
irvine